Topics/System_Design/behavioral_design_patterns.md

```diff
@@ -0,0 +1,122 @@
+# Behavioral Design Patterns
+
+In software engineering, behavioral design patterns  are a set of patterns that focus on the interaction between objects and how they communicate and collaborate with each other. These patterns address the assignment of responsibilities between objects, encapsulation of behavior, and the flow of control among them. They aim to improve the flexibility, extensibility, and maintainability of software systems by providing solutions to common communication and collaboration problems.
+
+## Table of Contents
+
+1. [Usefulness of Behavioral Design Patterns](#use)
+2. Examples of Behavioral Design Patterns
+    1. [Observer](#obs)
+    2. [Strategy](#str)
+    3. [Iterator](#itr)
+    4. [Command](#com)
+    5. Examples of other [Behavioral Design Patterns](https://refactoring.guru/design-patterns/behavioral-patterns)
+
+## Usage <a name="use"></a>
+
+Behavioral design patterns are useful when you need to define how objects interact in a flexible and reusable manner, ensuring that changes in one part of the system do not require modifications to other parts. They are particularly valuable in scenarios where the behavior of objects needs to vary dynamically during runtime. 
+
+Now, we will dicuss some commonly used behavioral patterns with real world analogies to get some intuition about when these patterns can be useful and how they can be implemented. For UML diagrams and implementation details check out [Refactoring Guru: Behavioral Patterns](https://refactoring.guru/design-patterns/behavioral-patterns). 
+
+## Observer Design Pattern <a name="obs"></a>
+
+The Observer design pattern is a behavioral pattern where an object, known as the subject, maintains a list of its dependents, called observers, and notifies them of any changes in its state. This pattern establishes a one-to-many relationship between the subject and its observers, allowing multiple objects to react to state changes in the subject independently. Observers register with the subject to receive notifications and update themselves accordingly. The Observer pattern promotes loose coupling between objects, making it easier to maintain and extend systems by separating concerns and allowing objects to interact without having direct knowledge of each other. It's commonly used in event handling, user interface design, and other scenarios where objects need to be notified of changes in another object's state.
+
+<p align="center">
+<img width="676" align="center" alt="Screenshot 2024-03-17 at 8 58 15 AM" src="https://github.com/paridhi26/CSC311Project/assets/85652265/f554584c-b2aa-4838-99b8-6a15c46310c7"> 
+</p>
+
+Image from [Refactoring Guru](https://refactoring.guru/design-patterns/observer)
```

Comment:
I think it would also be good to add in a small caption in addition to credit to the reference, to better aid what the image is trying to show.

Reply:
Added alt text for each image.